### DISCUSSION TITLE
Insider install problem (Win 11)

### DISCUSSION THREAD

**Discussion by tamasgt** [squidfunk/mkdocs-material#7333] [Q&A] [answered]
Hello,

something changed, everything was working, I tried to update the Material MkDocs Insider and got the following error:

![image](https://github.com/squidfunk/mkdocs-material/assets/5910815/9079dd0d-8c32-4dab-af99-5f5b7d5fdd63)

I tried updating python, pip, setuptools, but no matter what I do, it doesn't work. I tried the latest version of Python (3.12.3), now it's 3.11.9, but I can't install it. I tried:
```
pip        23.2.1
setuptools 65.5.0
```
or now with the last version

```
pip        24.1.1
setuptools 70.2.0
```
always the same error.
I cannot install either MkDocs free
```
pip install mkdocs-material
```
or insider
```
pip install --git+https://TOKEN@github.com/squidfunk/mkdocs-material-insiders.git@9.5.28-insiders-4.53.11
```
 in an empty python virtual environment. Always the same error. The regex throws the first error.

![image](https://github.com/squidfunk/mkdocs-material/assets/5910815/ed21fece-1eef-4d21-86c6-b6c630f87a6d)

Any idea what the problem might be?

Thanks,
Tamas

**Comment by kamilkrzyskow** [comment]
Hello @tamasgt,
the installation works for me on my normal Python 3.12 Windows 10 installation from python.org via the exe installer:
```
$ python -VV
Python 3.12.4 (tags/v3.12.4:8e8a4ba, Jun  6 2024, 19:30:16) [MSC v.1940 64 bit (AMD64)]
```
most likely, as the wheels for the `regex` module etc. are present for Windows and are compatible with my version.
https://pypi.org/project/regex/#files

However, when I try to run the install via MSYS2 and the mingw Python 3.11 I get the same crash:
```
$ python -VV
Python 3.11.9 (main, Apr 12 2024, 09:55:31)  [GCC 13.2.0 64 bit (AMD64)]
$ pip install mkdocs-material
# ...
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.mingw_x86_64-cpython-311
      creating build/lib.mingw_x86_64-cpython-311/regex
      copying regex_3/__init__.py -> build/lib.mingw_x86_64-cpython-311/regex
      copying regex_3/regex.py -> build/lib.mingw_x86_64-cpython-311/regex
      copying regex_3/_regex_core.py -> build/lib.mingw_x86_64-cpython-311/regex
      copying regex_3/test_regex.py -> build/lib.mingw_x86_64-cpython-311/regex
      running build_ext
      error: [WinError 2] The system cannot find the file specified
      [end of output]
```

I'm drawing a blank for now, I will have to investigate the build process.

~~cc @vedranmiletic @alexvoss Since you were the main contributors in https://github.com/squidfunk/mkdocs-material/pull/7163~~

**Reply by kamilkrzyskow** [reply]
OK, all I had to do was to install GCC in MSYS2 @tamasgt and the installation worked fine after that:
```
pacman -S mingw-w64-x86_64-gcc
```
I didn't actually get to the full traceback of the subprocess, just read over the `regex` `setup.py` and saw `.c` files in the extension modules, and the build crashed on the `build_ext` step... 
- https://github.com/mrabarnett/mrab-regex/blob/6d086ffc5be71220d527495ad384834e4f54f340/setup.py#L41-L42

Sorry for the mention guys 🫡

**Reply by tamasgt** [reply, marked as answer]
Thanks for the reply. I found the solution.

The problem was caused by the fact that [MSYS2](https://www.msys2.org/) in [documentation](https://squidfunk.github.io/mkdocs-material/plugins/requirements/image-processing/#cairo-graphics-windows) was installed (I installed it a long time ago.) and therefore the virtual environment is created differently by python. The solution was:
1) Uninstall MSYS2.
2) Uninstall Python
3) Delete directory C:\Users\USERNAME\AppData\Local\Programs\Python\PythonVERSION
3) Delete python virtual environment.
4) Install Python
5) Create python virtual environment.

and it works with latest Python version.
I will look at what you wrote #7163 and add if there is anything new. I'm still using this for now:
https://github.com/tschoonj/GTK-for-Windows-Runtime-Environment-Installer

Tamás

**Reply by kamilkrzyskow** [reply]
Hmm, you just uninstalled MSYS2 and installed the normal Python via the installer from python.org, right? 
Based on your steps you didn't reinstall MSYS2, I actually found the solution for the error 😁 
Good that you got it working anyways 👍

**Reply by vedranmiletic** [reply]
> Hmm, you just uninstalled MSYS2 and installed the normal Python via the installer from python.org, right? Based on your steps you didn't reinstall MSYS2, I actually found the solution for the error 😁 Good that you got it working anyways 👍

We should probably document the GCC requirement.

**Reply by kamilkrzyskow** [reply]
> We should probably document the GCC requirement.

Perhaps we should, but in general the Getting Started uses GCC as an example in step 6:
- https://www.msys2.org/

So I think the documentation here should be related to an error like the one above, where the wheel build fails, because of a lacking file.
`subprocess-exited-with-error`
`error: [WinError 2] The system cannot find the file specified`
etc.

**Reply by alexvoss** [reply]
When I worked on the documentation I was under the impression that GCC was getting installed by default. It seems sensible to me to add it to the list of things to install here rather than relying on the MSYS documentation. Need to fire up a VM to go and check it out tomorrow.

**Reply by tamasgt** [reply]
@alexvoss Don't waste your time. I looked at everything. The problem was that when image processing was first released, I also installed Python with MSYS2. (I don't have much experience with Python and MSYS2.)

There was an older project where the virtual environment was created before MSYS2 was installed, which worked. Now I created a new Material MkDocs project, but here the Python package installed with MSYS2 created the virtual environment. (And while I was looking for which version the problem was with, I got completely confused. I didn't understand why one worked and the other didn't.) After several hours, I noticed that the Python virtual environment of two projects was not the same. In the new project, Python installed with MSYS2 created the virtual environment, but it was nowhere to be seen. And I see this is what caused the installation error.

I have now removed everything. I installed Python using the Windows installer, MSYS2 and only the Cairo Graphics and pngquant packages. I'll test it now and report back.

**Reply by alexvoss** [reply]
Happy that the problem was solved for you but I could reproduce the issue with installing Insiders. Perhaps something I did way back when caused GCC to get installed. 

I spent a little time yesterday looking at the issue and I hope I can do a PR today that will decrease the chances of this occurring for others. I am not spending a lot of time on this, just taking my time as I mix it in with other activities.

**Reply by alexvoss** [reply]
After some experimentation, I have concluded that the instructions should probably be changed to use the MingW64 environment only. The default UCRT64 environment is a Posix emulation while MingW64 uses Windows APIs. I had no joy compiling Insiders in a UCRT64 environment. Using strictly MingW, installing Insiders works for me:

```
$ pacman -S mingw-w64-x86_64-python  mingw-w64-x86_64-python-pip
$ pacman -S mingw-w64-x86_64-gcc git
$ pip install git+https://${GH_TOKEN}@github.com/squidfunk/mkdocs-material-insiders.git
$ pacman -S mingw-w64-x86_64-pngquant
```

However, I am getting an error when trying to install Cairo:

```
$ pacman -S mingw-w64-x86_64-cairo
...
error: failed to commit transaction (conflicting files)
mingw-w64-x86_64-python-packaging: /mingw64/lib/python3.11/site-packages/packaging/__init__.py exists in filesystem
mingw-w64-x86_64-python-packaging: /mingw64/lib/python3.11/site-packages/packaging/__pycache__/__init__.cpython-311.pyc exists in filesystem
mingw-w64-x86_64-python-packaging: /mingw64/lib/python3.11/site-packages/packaging/__pycache__/_elffile.cpython-311.pyc exists in filesystem
...
```

It looks like this error occurs when installing the Insiders Edition before Cairo but not then installing the other way round. Anyone got an idea where this might be coming from?

Now, given that most people will want to use MSYS2 to install the libraries only and then use the Python version from python.org, I tested that and it works fine. Users do need to make sure that they have that version of Python in their path before any of the MSYS stuff. I can write this up as a PR but would like to hear what others think and perhaps observe on their systems.

P.S.: while deleting MSYS2 I notice that the whole environment is 108k files, which is a bit much for 2 libraries we actually want...

**Reply by tamasgt** [reply]
I think I found it.

The problem is that you install MSYS2 and when you create a Python virtual environment, it is created by the Python installed with MSYS2, and you cannot install Material for MkDocs in this virtual environment. (It works with the Windows Python installer.)

After my installation, in the Command Prompt window (not PowerShell):

![image](https://github.com/squidfunk/mkdocs-material/assets/5910815/4f2f671d-0f31-4f0a-870f-fe50c170b6c9)

**Solution 1**, this is how you create a Python virtual environment in project directory:

```PowerShell
C:\Users\USERNAME\AppData\Local\Programs\Python\PythonVERSION\python.exe -m venv customenviroment
```

**Solution 2**, Move the environment variables of the Python Windows installer, from user to system,

![2024-07-09_09-02-54](https://github.com/squidfunk/mkdocs-material/assets/5910815/11d1cfc7-1b1d-488a-aff0-cc57c5bfcd90)

and in this order:

![2024-07-09_09-11-29](https://github.com/squidfunk/mkdocs-material/assets/5910815/56878306-ec04-4114-9aa6-99edb167c3b4)

Restart terminal, the result, we changed the order:

![2024-07-09_09-21-02](https://github.com/squidfunk/mkdocs-material/assets/5910815/dbf212a1-d7f2-47f5-85d9-fe126fa0d02f)

In the project directory:

```PowerShell
python -m venv customenviroment
```

And the virtual environment is created by the Windows installer of Python. And not the MSYS2 installer of Python. And everything seems to work.

**Reply by alexvoss** [reply]
Yep, the choreography with multiple shells (CMD, Powershell, MSYS2) and environments with MSYS2 (there are 6 if you exclude the aarch64 one) as well as possible multiple installations of Python all make me hesitate to actually put together the PR. We may never be able to describe this all adequately but need to find a version that works in a large enough majority of cases. 

I fear we may have opened a pandora's box with the suggestion to install the dependencies via MSYS2. It is super easy when starting from scratch and following instructions to the letter but deviate and you are doomed. Windows is fickle anyway. We had to consider "Roaming Profiles" recently, a thing I never even wanted to know about... Ok, so Homebrew on Mac is also not trivial once you manage to mess it up and use the built-in Python from MacOS. Windows, however, is an order of magnitude more fickle.

**Reply by tamasgt** [reply]
And the social card and image optimization works with the MSYS2 packages.

**Reply by tamasgt** [reply]
I think the documentation of the Material for MkDocs MSYS2 packages is now good, it was installed in a few minutes. The problem is that you don't even know that another Python is installed, the order of the environment variables changes, and the same command starts using another Python program. I don't know what the good solution is in the documentation.

I think everyone starts with the Windows installer of Python, if you install MSYS afterwards, you will definitely have this problem. I saw that the installation problem would only occur if you were also using a Python virtual environment. Because the created virtual environment is the wrong one.

An interesting question is, if MSYS is installed this way, does the user update the Windows installer of Python in vain, because he does not even use it due to changed environment variables. :smiley:

**Reply by vedranmiletic** [reply]
> And the social card and image optimization works with the MSYS2 packages.

Yes, that's why it is recommended to use it. Unlike installers for Windows, its packages are kept up to date and can be relied upon.

> An interesting question is, if MSYS is installed this way, does the user update the Windows installer of Python in vain, because he does not even use it due to changed environment variables. 😃

If the user is aware of it and changes environment variable accordingly, it is not in vain. But good point, it should be noted in the documentation.
